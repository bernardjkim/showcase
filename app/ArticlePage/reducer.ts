/*
 *
 * ArticlePage reducer
 *
 */

import { ArticleActionTypes, ArticleState, SET_ARTICLE_ID } from './types';

// The initial state of the App
export const initialState: ArticleState = {
  articleId: undefined,
};

function articlePageReducer(state: ArticleState = initialState, action: ArticleActionTypes) {
  switch (action.type) {
    case SET_ARTICLE_ID:
      return {
        ...state,
        articleId: action.articleId,
      };

    default:
      return state;
  }
}

export default articlePageReducer;
