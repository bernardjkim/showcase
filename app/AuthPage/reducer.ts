/*
 *
 * AuthPage reducer
 *
 */

// import { LOCATION_CHANGE } from 'react-router-redux';
import {
  AuthState,
  AuthActionTypes,
  CLEAR_ERRORS,
  CREATE_TOKEN,
  CREATE_TOKEN_ERROR,
  CREATE_TOKEN_SUCCESS,
  CREATE_USER,
  CREATE_USER_ERROR,
  CREATE_USER_SUCCESS,
} from './types';

export const initialState: AuthState = {
  loading: false,
  error: undefined,
};

function AuthPageReducer(
  state: AuthState = initialState,
  action: AuthActionTypes,
) {
  switch (action.type) {
    // case LOCATION_CHANGE:
    //   return { ...state, error: undefined };

    case CLEAR_ERRORS:
      return { ...AbortController, error: undefined };

    case CREATE_TOKEN:
      return { ...state, loading: true, error: undefined };

    case CREATE_TOKEN_SUCCESS:
      return { ...state, loading: false };

    case CREATE_TOKEN_ERROR: {
      return { ...state, loading: false, error: { createToken: action.error } };
    }

    case CREATE_USER:
      return { ...state, loading: true, error: undefined };

    case CREATE_USER_SUCCESS:
      return { ...state, loading: false };

    case CREATE_USER_ERROR:
      return { ...state, loading: false, error: { createUser: action.error } };

    default:
      return state;
  }
}

export default AuthPageReducer;
