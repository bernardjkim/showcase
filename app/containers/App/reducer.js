/*
 * 
 * App reducer
 * 
 */

import { fromJS } from 'immutable';
import {
  CREATE_TOKEN,
  CREATE_TOKEN_ERROR,
  CREATE_TOKEN_SUCCESS,
} from './constants';

// The initial state of the App
export const initialState = fromJS({
  loading: false,
  error: false,
  token: false,
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case CREATE_TOKEN:
      return state.set('loading', true).set('error', false);

    case CREATE_TOKEN_SUCCESS:
      return state.set('token', action.token).set('loading', false);

    case CREATE_TOKEN_ERROR:
      return state.set('error', fromJS(action.error)).set('loading', false);

    default:
      return state;
  }
}

export default appReducer;
