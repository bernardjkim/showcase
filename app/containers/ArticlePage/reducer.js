/*
 *
 * ArticlePage reducer
 *
 */

import { fromJS } from 'immutable';
import {
  CREATE_COMMENT,
  CREATE_COMMENT_ERROR,
  CREATE_COMMENT_SUCCESS,
  LIKE_ARTICLE,
  LIKE_ARTICLE_ERROR,
  LIKE_ARTICLE_SUCCESS,
  LOAD_ARTICLE,
  LOAD_ARTICLE_ERROR,
  LOAD_ARTICLE_SUCCESS,
  LOAD_COMMENTS,
  LOAD_COMMENTS_ERROR,
  LOAD_COMMENTS_SUCCESS,
} from './constants';

// The initial state of the App
export const initialState = fromJS({
  loading: false,
  error: false,
  /**
   * article
   * @field {string}  _id         - Article id
   * @field {string}  uri         - Article uri
   * @field {string}  github      - GitHub repo
   * @field {string}  description - Article description
   * @field {string}  image       - Article image link
   * @field {number}  likes       - Number of likes
   * @field {bool}    likedByUser - Article is liked by user
   * @field {array}   comments    - Array of comments
   */
  article: false,
});

function articlePageReducer(state = initialState, action) {
  switch (action.type) {
    case CREATE_COMMENT:
      return state.set('loading', true).set('error', false);

    case CREATE_COMMENT_SUCCESS:
      return state
        .updateIn(['article', 'comments'], comments =>
          comments.push(fromJS(action.comment)),
        )
        .set('loading', false);

    case CREATE_COMMENT_ERROR:
      return state.set('error', fromJS(action.error)).set('loading', false);

    case LIKE_ARTICLE:
      return state.set('loading', true).set('error', false);

    case LIKE_ARTICLE_SUCCESS:
      return state
        .updateIn(['article', 'likes'], likes => likes + 1)
        .setIn(['article', 'likedByUser'], true)
        .set('loading', false);

    case LIKE_ARTICLE_ERROR:
      return state.set('error', fromJS(action.error)).set('loading', false);

    case LOAD_ARTICLE:
      return state
        .set('loading', true)
        .set('error', false)
        .set('article', false);

    case LOAD_ARTICLE_SUCCESS:
      return state.set('article', fromJS(action.article)).set('loading', false);

    case LOAD_ARTICLE_ERROR:
      return state.set('error', fromJS(action.error)).set('loading', false);

    case LOAD_COMMENTS:
      return state.set('loading', true).set('error', false);

    case LOAD_COMMENTS_SUCCESS:
      return state.setIn(['article', 'comments'], fromJS(action.comments));

    case LOAD_COMMENTS_ERROR:
      return state.set('error', fromJS(action.error)).set('loading', false);

    default:
      return state;
  }
}

export default articlePageReducer;
