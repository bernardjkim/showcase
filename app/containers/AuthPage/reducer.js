/*
 *
 * AuthPage reducer
 *
 */

import { fromJS } from 'immutable';
import { LOCATION_CHANGE } from 'react-router-redux';
import {
  CLEAR_ERRORS,
  CREATE_TOKEN,
  CREATE_TOKEN_ERROR,
  CREATE_TOKEN_SUCCESS,
  CREATE_USER,
  CREATE_USER_ERROR,
  CREATE_USER_SUCCESS,
} from './constants';

export const initialState = fromJS({
  loading: false,
  error: false,
});

function AuthPageReducer(state = initialState, action) {
  switch (action.type) {
    case LOCATION_CHANGE:
      return state.set('error', false);

    case CLEAR_ERRORS:
      return state.set('error', false);

    case CREATE_TOKEN:
      return state.set('loading', true).set('error', false);

    case CREATE_TOKEN_SUCCESS:
      return state.set('loading', false);

    case CREATE_TOKEN_ERROR: {
      return state
        .set('error', fromJS({ createToken: action.error }))
        .set('loading', false);
    }

    case CREATE_USER:
      return state.set('loading', true).set('error', false);

    case CREATE_USER_SUCCESS:
      return state.set('loading', false);

    case CREATE_USER_ERROR:
      return state
        .set('error', fromJS({ createUser: action.error }))
        .set('loading', false);

    default:
      return state;
  }
}

export default AuthPageReducer;
