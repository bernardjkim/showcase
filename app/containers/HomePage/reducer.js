/*
 *
 * HomePage reducer
 *
 */

import { fromJS } from 'immutable';
import {
  LIKE_ARTICLE,
  LIKE_ARTICLE_ERROR,
  LIKE_ARTICLE_SUCCESS,
  LOAD_ARTICLE,
  LOAD_ARTICLE_ERROR,
  LOAD_ARTICLE_SUCCESS,
} from './constants';

// The initial state of the App
export const initialState = fromJS({
  loading: false,
  error: false,
  article: false,
});

function homePageReducer(state = initialState, action) {
  switch (action.type) {
    case LIKE_ARTICLE:
      return state.set('loading', true).set('error', false);

    case LIKE_ARTICLE_SUCCESS: {
      const article = state.get('article');
      article.likes += 1;
      return state.set('article', article).set('loading', false);
    }
    case LIKE_ARTICLE_ERROR:
      return state.set('error', action.error).set('loading', false);

    case LOAD_ARTICLE:
      return state
        .set('loading', true)
        .set('error', false)
        .set('article', false);

    case LOAD_ARTICLE_SUCCESS:
      return state.set('article', action.article).set('loading', false);

    case LOAD_ARTICLE_ERROR:
      return state.set('error', action.error).set('loading', false);

    default:
      return state;
  }
}

export default homePageReducer;
