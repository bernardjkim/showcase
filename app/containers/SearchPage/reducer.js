/*
 *
 * SearchPage reducer
 *
 */

import { fromJS } from 'immutable';
import {
  CLEAR_STATE,
  SET_SEARCH,
  LOAD_ARTICLES,
  LOAD_ARTICLES_ERROR,
  LOAD_ARTICLES_SUCCESS,
  LOAD_NEXT,
  LOAD_NEXT_ERROR,
  LOAD_NEXT_SUCCESS,
} from './constants';

export const initialState = fromJS({
  loading: false,
  error: false,
  articles: [],
  search: '',
  offset: 0,
});

function searchPageReducer(state = initialState, action) {
  switch (action.type) {
    case CLEAR_STATE:
      return initialState;

    case SET_SEARCH:
      return state.set('search', action.search);

    case LOAD_ARTICLES:
      return state
        .set('articles', fromJS([]))
        .set('offset', 0)
        .set('loading', true)
        .set('error', false);

    case LOAD_ARTICLES_SUCCESS:
      return state
        .set('articles', fromJS(action.articles))
        .set('offset', action.articles.length)
        .set('loading', false);

    case LOAD_ARTICLES_ERROR:
      return state.set('error', fromJS(action.error)).set('loading', false);

    case LOAD_NEXT:
      return state.set('loading', true).set('error', false);

    case LOAD_NEXT_SUCCESS:
      return state
        .update('articles', articles =>
          articles.concat(fromJS(action.articles)),
        )
        .update('offset', val => val + action.articles.length)
        .set('loading', false);

    case LOAD_NEXT_ERROR:
      return state.set('error', fromJS(action.error)).set('loading', false);

    default:
      return state;
  }
}

export default searchPageReducer;
