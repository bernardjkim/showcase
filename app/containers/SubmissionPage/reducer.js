/*
 *
 * SubmissionPage reducer
 *
 */

import { fromJS } from 'immutable';
import {
  SUBMIT_FORM,
  SUBMIT_FORM_ERROR,
  SUBMIT_FORM_SUCCESS,
} from './constants';

export const initialState = fromJS({
  loading: false,
  error: false,
  submissionSuccess: false,
});

function submissionPageReducer(state = initialState, action) {
  switch (action.type) {
    case SUBMIT_FORM:
      return state
        .set('submissionSuccess', false)
        .set('loading', true)
        .set('error', false);

    case SUBMIT_FORM_SUCCESS:
      return state.set('submissionSuccess', true).set('loading', false);

    case SUBMIT_FORM_ERROR:
      return state
        .set('submissionSuccess', false)
        .set('error', fromJS(action.error))
        .set('loading', false);

    default:
      return state;
  }
}

export default submissionPageReducer;
