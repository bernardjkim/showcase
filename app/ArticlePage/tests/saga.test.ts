/**
 * Test sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { cloneableGenerator } from '@redux-saga/testing-utils';
import { all, put, takeLatest } from 'redux-saga/effects';

import {
  createComment as createCommentAction,
  createCommentError,
  createCommentSuccess,
  likeArticleError,
  likeArticleSuccess,
  loadArticle as loadArticleAction,
  loadArticleError,
  loadArticleSuccess,
  loadCommentsError,
  loadCommentsSuccess,
  loadLikesError,
  loadLikesSuccess,
} from '../actions';
import { initialState } from '../reducer';
import articlePageSaga, { createComment, likeArticle, loadArticle, loadComments, loadLikes } from '../saga';
import {
  ArticleState,
  Comment,
  CreateCommentAction,
  CREATE_COMMENT,
  Like,
  LoadArticleAction,
  LIKE_ARTICLE,
  LOAD_ARTICLE,
  LOAD_COMMENTS,
  LOAD_LIKES,
} from '../types';

const state: ArticleState = initialState;

const article = {
  _id: '_id',
  title: 'title',
  uri: 'uri',
  github: 'github',
  description: 'description',
  tags: [],
  image: 'image',
};
const comments: Comment[] = [];
const likes: Like[] = [];
const error = new Error('Test Error');

state.article = article;

describe('createComment Saga', () => {
  const comment: Comment = {};
  const gen = cloneableGenerator(createComment)(createCommentAction(comment) as CreateCommentAction);

  const selectDescriptor = gen.next().value;
  expect(selectDescriptor).toMatchSnapshot();

  const callDescriptor = gen.next(state.article).value;
  expect(callDescriptor).toMatchSnapshot();

  it('should dispatch the createCommentSuccess action if successful', () => {
    const putDescriptor = gen.clone().next({ comment }).value;
    expect(putDescriptor).toEqual(put(createCommentSuccess(comment)));
  });

  it('should dispatch the createCommentError action if error', () => {
    const clone = gen.clone();
    const putDescriptor = clone.throw ? clone.throw(error).value : null;
    expect(putDescriptor).toEqual(put(createCommentError(error)));
  });
});

describe('likeArticle Saga', () => {
  const like = {};
  const gen = cloneableGenerator(likeArticle)();

  const selectDescriptor = gen.next().value;
  expect(selectDescriptor).toMatchSnapshot();

  const callDescriptor = gen.next(state.article).value;
  expect(callDescriptor).toMatchSnapshot();

  it('should dispatch the likeArticleSuccess action if successful', () => {
    const putDescriptor = gen.clone().next({ like }).value;
    expect(putDescriptor).toEqual(put(likeArticleSuccess(like)));
  });

  it('should dispatch the likeArticleError action if error', () => {
    const clone = gen.clone();
    const putDescriptor = clone.throw ? clone.throw(error).value : null;
    expect(putDescriptor).toEqual(put(likeArticleError(error)));
  });
});

describe('loadArticle Saga', () => {
  const query = { id: 'id' };
  const gen = cloneableGenerator(loadArticle)(loadArticleAction(query) as LoadArticleAction);

  // const selectDescriptor = gen.next().value;
  // expect(selectDescriptor).toMatchSnapshot();

  const callDescriptor = gen.next().value;
  expect(callDescriptor).toMatchSnapshot();

  it('should dispatch the loadArticleSuccess action if successful', () => {
    const putDescriptor = gen.clone().next({ article }).value;
    expect(putDescriptor).toEqual(put(loadArticleSuccess(article)));
  });

  it('should dispatch the loadArticleError action if error', () => {
    const clone = gen.clone();
    const putDescriptor = clone.throw ? clone.throw(error).value : null;
    expect(putDescriptor).toEqual(put(loadArticleError(error)));
  });
});

describe('loadComments Saga', () => {
  const gen = cloneableGenerator(loadComments)();

  const selectDescriptor = gen.next().value;
  expect(selectDescriptor).toMatchSnapshot();

  const callDescriptor = gen.next(state.article).value;
  expect(callDescriptor).toMatchSnapshot();

  it('should dispatch the loadCommentsSuccess action if successful', () => {
    const putDescriptor = gen.clone().next({ comments }).value;
    expect(putDescriptor).toEqual(put(loadCommentsSuccess(comments)));
  });

  it('should dispatch the loadCommentsError action if error', () => {
    const clone = gen.clone();
    const putDescriptor = clone.throw ? clone.throw(error).value : null;
    expect(putDescriptor).toEqual(put(loadCommentsError(error)));
  });
});

describe('loadLikes Saga', () => {
  const gen = cloneableGenerator(loadLikes)();

  const selectDescriptor = gen.next().value;
  expect(selectDescriptor).toMatchSnapshot();

  const callDescriptor = gen.next(state.article).value;
  expect(callDescriptor).toMatchSnapshot();

  it('should dispatch the loadLikesSuccess action if successful', () => {
    const putDescriptor = gen.clone().next({ likes }).value;
    expect(putDescriptor).toEqual(put(loadLikesSuccess(likes)));
  });

  it('should dispatch the loadLikesError action if error', () => {
    const clone = gen.clone();
    const putDescriptor = clone.throw ? clone.throw(error).value : null;
    expect(putDescriptor).toEqual(put(loadLikesError(error)));
  });
});

describe('articlePageSaga Saga', () => {
  const saga = articlePageSaga();

  it('should start task to watch for api request action', () => {
    const takeLatestDescriptor = saga.next().value;
    expect(takeLatestDescriptor).toEqual(
      all([
        takeLatest(CREATE_COMMENT, createComment),
        takeLatest(LIKE_ARTICLE, likeArticle),
        takeLatest(LOAD_ARTICLE, loadArticle),
        takeLatest(LOAD_COMMENTS, loadComments),
        takeLatest(LOAD_LIKES, loadLikes),
      ]),
    );
  });
});
