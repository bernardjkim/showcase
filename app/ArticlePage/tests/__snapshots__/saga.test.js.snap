// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[` 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/comment",
      Object {
        "body": "articleId=5c4ebe9be5346639c09786e3",
        "headers": Object {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[` 3`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[` 4`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/like",
      Object {
        "body": "articleId=5c4ebe9be5346639c09786e3",
        "headers": Object {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[` 5`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/article/id",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[` 6`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[` 7`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/comment/?article=5c4ebe9be5346639c09786e3",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[` 8`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[` 9`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/api/like/?article=5c4ebe9be5346639c09786e3",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
